utils/polus-filepattern-util/filepattern/classes.py

```diff
@@ -0,0 +1,269 @@
+        """Determine an output name for a list of files
+        
+        See the :any:`output_name` method for more details.
+        
+        This method uses the ``filepattern`` used to initialize the object to
+        determine an output file name that summarizes the range of variables
+        included in the ``file_path`` list of dictionaries. If ``file_path`` is
+        empty, this method returns an output file name that summarizes the range
+        of all variables parsed by the object.
+
+        Args:
+            files: A list of file dictionaries
+
+        Returns:
+            An output file name
+        """
+        
+        if len(files) == 0:
+            files = self.files
+            files = get_matching(files,self.var_order,**{k.upper():v for k,v in self.uniques.items()})
+        # else:
+        #     files,uniques = _parse(parse_filename,
+        #                            [file['file'] for file in files],
+        #                            self.pattern,
+        #                            self.regex,
+        #                            self.variables,
+        #                            self.var_order)
+        
+        vals = {v:set() for v in self.var_order}
+        for file in files:
+            for k,v in file.items():
+                if k not in self.var_order:
+                    continue
+                vals[k].add(v)
+        
+        kwargs = {}
+        for k,v in vals.items():
+            v = list(v)
+            if len(v) == 1 and v[0] != -1:
+                kwargs[k] = v[0]
+                
+        return output_name(self.pattern,files,kwargs)
+
+    # Get filenames matching values for specified variables
+    def get_matching(self,**kwargs):
+        """ Get all filenames matching specific values
+    
+        This function runs the get_matching function using the objects file
+        dictionary. For more information, see :any:`get_matching`.
+
+        Args:
+            **kwargs: One of :any:`VARIABLES`, must be uppercase, can be single
+                values or a list of values
+                
+        Returns:
+            A list of all files matching the input values
+        """
+        # get matching files
+        files = get_matching(self.files,self.var_order,out_var=None,**kwargs)
+        return files
+
+    def __iter__(self):
+        
+        group_by = self._group_by
+        kwargs = self._kwargs
+        self._group_by = None
+        self._kwargs = None
+        
+        if kwargs == None:
+            raise SyntaxError('Cannot directly iterate over a FilePattern object. Call it (i.e. for i in fp())')
+
+        # If self.files is a list, no parsing took place so just loop through the files
+        if isinstance(self.files,list):
+            for f in self.files:
+                yield [f]
+            return
+
+        # Generate the values to iterate through
+        iter_vars = {}
+        for v in self.var_order:
```

Comment:
This loop is not very readable, especially considering you're storing multiple data types in your dictionary.  A few oneliner comments explainign what's going on might be useful, and spare you (or someone) else the headache later on down the road of trying to figure out what's going on. The same goes for the second loop at the check for `-1` in `iter_vars` which requires knowledge of what happened earlier to understand.

Reply:
I added a lot of comments to the loops. Hopefully this helps clear up the confusion. Looking back at it, I see how confusing this is to follow.